### PR TITLE
#3510 Defer logging-related warnings until after the startup indicator

### LINKED ISSUE #3510: Defer logging-related warnings/errors until after initial startup message
Some of the new logging-related warnings appear prior to the 'Sync Gateway startup' log message, e.g.:

```
Adams-MacBook-Pro-3:bin adam$ ./sync_gateway ~/config/default.json
2018-04-25T11:33:20.999-07:00 [WRN] Using deprecated config option: log. Use logging.console.log_keys instead. -- rest.(*ServerConfig).deprecatedConfigLoggingFallback() at config.go:607
2018-04-25T11:33:20.999-07:00 [WRN] Deprecated log key: "Changes+" found. Changing to: "Changes". -- base.ToLogKey() at log_keys.go:172
2018-04-25T11:33:20.999-07:00 [ERR] No logFilePath configured, and --defaultLogFilePath flag is not set. Log files required for product support are not being generated. -- base.(*LoggingConfig).Init() at logging_config.go:40
2018-04-25T11:33:20.999-07:00 ==== Couchbase Sync Gateway/feature/issue_accel_203(fac51dc+CHANGES) ====
```

Since support (and ourselves) are used to using the `==== Couchbase Sync Gateway` message to identify SG restarts when parsing logs, it would be useful to defer the log warnings/messages if possible.

### PR DESCRIPTION
**Pull request by bbrks**
Fixes #3510 

- Defers logging-related warnings/errors until after the startup indicator has been logged
- Removed `Broadcastf` and replaced with `LogSyncGatewayVersion`
- Re-add `--verbose` flag check to enable `HTTP+` log key

## Before/After
screen shot 2018-05-16 at 15 53 47

### CODE REVIEW

File: base/log_keys.go [hunk 1]
```diff
@@ -142,8 +142,10 @@ func (keyMask *LogKey) EnabledLogKeys() []string {
 }
 
 // ToLogKey takes a slice of case-sensitive log key names and will return a LogKey bitfield.
-func ToLogKey(keysStr []string) LogKey {
+func ToLogKey(keysStr []string) (LogKey, []DeferredLog) {
```
Comment (tleyden) [inline base/log_keys.go]: Can you update the comments to describe what the purpose of the `[]DeferredLog` return value is?

File: base/log_keys.go [hunk 2]
```diff
@@ -155,17 +157,28 @@ func ToLogKey(keysStr []string) LogKey {
 		// Strip a single "+" suffix in log keys and warn (for backwards compatibility)
 		if strings.HasSuffix(name, "+") {
 			newName := strings.TrimSuffix(name, "+")
-			Warnf(KeyAll, "Deprecated log key: %q found. Changing to: %q.", name, newName)
+
+			// Need to take a copy of name, as the loop will mutate the value
+			nameCpy := name
```
Comment (tleyden) [inline base/log_keys.go]: Are you sure this is actually needed since [Go strings are immutable](https://groups.google.com/d/msg/golang-nuts/naMCI9Jt6Qg/4QUebtQz0YAJ)?  I think the later statement `name = newName` shouldn't affect adding a closure function to `deferredLogs`

File: base/logging.go [hunk 3]
```diff
@@ -34,6 +34,9 @@ const (
 
 type Level int32
 
+// DeferredLog is a log function that can be logged at a later date.
+type DeferredLog func()
```
Comment (tleyden) [inline base/logging.go]: What about changing this to `DeferredLogFn` to highlight that it's a function rather than a log message/statement?   On my first reading of `var deferredLogs []DeferredLog` I had no idea that was a slice of functions.

File: base/logging_config.go [hunk 4]
```diff
@@ -25,48 +28,47 @@ type LoggingConfig struct {
...
 	if !hasLogFilePath(&c.LogFilePath, defaultLogFilePath) {
-		Errorf(KeyAll, "No logFilePath configured, and --defaultLogFilePath flag is not set. Log files required for product support are not being generated.")
-		return nil
+		return deferredLogs, ErrUnsetLogFilePath
 	}
 
 	err = validateLogFilePath(&c.LogFilePath, defaultLogFilePath)
 	if err != nil {
-		return err
+		return
```
Comment (tleyden) [inline base/logging_config.go]: I personally prefer "non-naked returns", even if they're redundant, although I realize this might be more idiomatic go.  Taking a step back, I feel like as a team it would be great to have an agreed "coding style guide" document we follow, a la https://github.com/golang/go/wiki/CodeReviewComments

File: rest/config.go [hunk 5]
```diff
@@ -564,65 +567,87 @@ func (config *ServerConfig) setupAndValidateDatabases() error {
...
+	return deferredLogs, nil
 }
 
 // deprecatedConfigLoggingFallback will parse the ServerConfig and try to
 // use older logging config options for backwards compatibility.
-func (config *ServerConfig) deprecatedConfigLoggingFallback(verbose bool) {
+// It will return a slice of deferred log functions to execute at a later time.
+func (config *ServerConfig) deprecatedConfigLoggingFallback() []base.DeferredLog {
+
+	deferredLogs := make([]base.DeferredLog, 0)
```
Comment (tleyden) [inline rest/config.go]: Another style guide issue .. should this be `make(..., 0)` or `deferredLogs := []base.DeferredLog{}`?  I'm fine with either, but I guess it would be good to have a consistent approach.

File: rest/config.go [hunk 6]
```diff
@@ -564,65 +567,87 @@ func (config *ServerConfig) setupAndValidateDatabases() error {
...
 
 	if config.Logging.DeprecatedDefaultLog != nil {
 		// Fall back to the old logging.["default"].LogFilePath option
 		if config.Logging.LogFilePath == "" && config.Logging.DeprecatedDefaultLog.LogFilePath != nil {
-			base.Warnf(base.KeyAll, "Using deprecated config option: logging.[\"default\"].LogFilePath. Use logging.log_file_path instead.")
+			deferredLogs = append(deferredLogs, func() {
+				base.Warnf(base.KeyAll, "Using deprecated config option: logging.[\"default\"].LogFilePath. Use logging.log_file_path instead.")
```
Comment (tleyden) [inline rest/config.go]: What about using a backtick surround string to avoid having to escape the double quotes?

File: rest/config.go [hunk 7]
```diff
@@ -564,65 +567,87 @@ func (config *ServerConfig) setupAndValidateDatabases() error {
...
+				base.Warnf(base.KeyAll, "Using deprecated config option: logging.[\"default\"].LogKeys. Use logging.console.log_keys instead.")
+			})
 			config.Logging.Console.LogKeys = config.Logging.DeprecatedDefaultLog.LogKeys
 		}
 
 		// Fall back to the old logging.["default"].LogLevel option
 		if config.Logging.Console.LogLevel == nil && config.Logging.DeprecatedDefaultLog.LogLevel != 0 {
-			base.Warnf(base.KeyAll, "Using deprecated config option: logging.[\"default\"].LogLevel. Use logging.console.log_level instead.")
+			deferredLogs = append(deferredLogs, func() {
+				base.Warnf(base.KeyAll, "Using deprecated config option: logging.[\"default\"].LogLevel. Use logging.console.log_level instead.")
```
Comment (tleyden) [inline rest/config.go]: Since this string is repeated, might be good to pull out into a variable.

File: rest/config.go [hunk 8]
```diff
@@ -760,6 +785,15 @@ func ParseCommandLine(runMode SyncGatewayRunMode) {
 			config.SkipRunmodeValidation = *skipRunModeValidation
 		}
 
+		if defaultLogFilePathFlag != nil {
+			defaultLogFilePath = *defaultLogFilePathFlag
+		}
+
+		// Log HTTP Responses if verbose is enabled.
+		if verbose != nil && *verbose {
```
Comment (tleyden) [inline rest/config.go:802]: This seems a bit unrelated to this PR.. was it supposed to sneak in?  Also, I wish we could just remove this option and let people specify `HTTP+` directly, but that's probably a different conversation.

File: rest/config.go [hunk 9]
```diff
@@ -959,6 +984,24 @@ func ValidateConfigOrPanic(runMode SyncGatewayRunMode) {
...
+	// or from a sync_gateway config file so we can validate the
+	// configuration and setup logging now
+	deferredLogs, err := config.setupAndValidateLogging()
+	if err != nil {
+		base.Fatalf(base.KeyAll, "Error setting up logging: %v", err)
+	}
+
+	// This is the earliest opportunity to print a startup indicator
+	// that will be persisted in log files
+	base.Broadcastf("==== %s ====", base.LongVersionString)
```
Comment (tleyden) [inline rest/config.go]: Broadcast sounds like it's broadcasting over a network.  I think this would be clearer as `LogToAllTargets()` or `LogToAllDests()`

**Review by tleyden — CHANGES_REQUESTED** [review]
The overall approach feels more complicated than it should be .. though I don't have another suggested approach.  I'd be very curious to see what standardized logging frameworks like logrus do to "bootstrap" their logging services.

File: base/log_keys.go (hunk 2 above)
Comment (bbrks) [reply base/log_keys.go]: Probably a poorly written comment on my part, as I don't think it's actually mutating name, rather replacing the reference to it. I could add an argument to the closure/anonymous function, but that would require changing the signature of `DeferredLog`.

Assigning a new variable is the other suggested solution in https://golang.org/doc/faq#closures_and_goroutines

>  Even easier is just to create a new variable, using a declaration style that may seem odd but works fine in Go:
> ```go
> for _, v := range values {
>         v := v // create a new 'v'.
>         go func() {
>             fmt.Println(v)
>             done <- true
>         }()
>     }
> ```

File: base/logging_config.go (hunk 4 above)
Comment (bbrks) [reply base/logging_config.go]: Agreed, I dislike naked returns, will restore this.

File: rest/config.go (hunk 8 above)
Comment (bbrks) [reply rest/config.go:802]: I think I noticed it was a regression in behavior for 2.1 as I was writing this PR. (i.e. --verbose was being ignored after the new logging went in)

I can split it out if necessary.

File: base/log_keys.go (hunk 2 above)
Comment (tleyden) [reply base/log_keys.go]: But I think that's only relevant when there's another goroutine in play.  Here I didn't see one .. although maybe I didn't trace back far  enough.

Comment (bbrks) [reply base/log_keys.go]: It can occur when goroutines are used, but more generally, when closures are executed after the loop as exited.

File: rest/config.go (hunk 9 above)
Comment (bbrks) [reply rest/config.go]: It was difficult naming this when coming up with the logging API. In practice, this is the *only* place it is ever used, and should perhaps be removed in exchange for `base.PrintVersion()`.

The intention with `Broadcastf` was to cause second thought when writing it, as it circumvents a few user-defined log settings (i.e. `log_level=none` will still log these app startup indicators)

**Comment by bbrks** [comment]
@tleyden I suspect this is a bit different from normal logging frameworks, as we have a requirement to log something (the ==== message) at the startup of the application, and have it always be the first log emitted. We can't log it before setting up logging, as it would never end up in the files we want.

Log frameworks have the advantage of relying pretty much only on stdout/stderr, or one log file, whereas we have 5 different outputs that may or may not be enabled.

My first attempt was to start up logging in a "paused" state, and then release logs once it'd finished initializing, but was deemed to general-purpose/dangerous, so the log-setup code specifically was changed to feed back deferred logs.

File: base/log_keys.go (hunk 2 above)
Comment (tleyden) [reply base/log_keys.go]: Here's a way to refactor the code to eliminate the need for copying that value to workaround the closure scope issue.   (I started to write a description but ended up sketching it out in some code)

https://github.com/couchbase/sync_gateway/blob/feature/issue_3510_remove_var_copy/base/log_keys.go#L144-L193

I verified that the test does still pass.

Threw in some function / variable renaming suggestions too.. but don't feel all that strongly about those, so if you like this change feel free to cherry pick the parts you like and skip the renaming changes.

File: rest/config.go (hunk 8 above)
Comment (tleyden) [reply rest/config.go:802]: Probably not needed, I was mainly making sure it was intended to be included.

File: rest/config.go (hunk 9 above)
Comment (tleyden) [reply rest/config.go]: Yeah I'd personally prefer it to be moved to `base.LogSyncGatewayVersion()` or `base.PrintSyncGatewayVersion()` over `base.Broadcastf()`.  "Broadcast" has different connotations .. like in networking or even condition variables for thread sync.  If we had to keep "Broadcast", maybe "LogBroadcast" would at least be slightly better.

**Comment by tleyden** [comment]
> Log frameworks have the advantage of relying pretty much only on stdout/stderr, or one log file, whereas we have 5 different outputs that may or may not be enabled.

Ok it sounds like you've thought through the alternatives, and like I said I don't know of a better alternative.. just wanted to make sure we're checking prior art.

**Review by tleyden — APPROVED** [review]

**Review by adamcfraser — APPROVED** [review]
